### PR TITLE
Add support for automatic reloading the config when it has been changed (based on #50)

### PR DESCRIPTION
**Pull request by marcusramberg**
Automatic reloading is very useful, and I've been using it locally, but #50 seems to be abandoned, so I forked it and tried to address the remaining review comments.

### CODE REVIEW

File: deck.go [hunk 1]
```diff
@@ -70,6 +72,42 @@ func LoadDeck(dev *streamdeck.Device, base string, deck string) (*Deck, error) {
 		d.Widgets = append(d.Widgets, w)
 	}
 
+	watcher, err := fsnotify.NewWatcher()
+	if err == nil {
+		err = watcher.Add(path)
+		if err == nil {
+
+			go func() {
```
Comment (muesli) [inline deck.go:80]: The problem with this is that we'll keep spawning new watchers every time we load a deck, even if it's one that has been loaded previously. We'll either need a central monitor that watches decks and remembers all the existing watchers or a way to terminate previous launched watchers.

File: deck.go [hunk 2]
```diff
@@ -70,6 +72,42 @@ func LoadDeck(dev *streamdeck.Device, base string, deck string) (*Deck, error) {
...
+				for {
+					select {
+					case event := <-watcher.Events:
+						if currentDeck == path {
+							fmt.Printf("Change:  %s: %s\n", event.Op, event.Name)
```
Comment (muesli) [inline deck.go:85]: Spurious space here between `Change:` and `%s`.

File: deck.go [hunk 3]
```diff
@@ -70,6 +72,42 @@ func LoadDeck(dev *streamdeck.Device, base string, deck string) (*Deck, error) {
...
+							if err != nil {
+								fatal(err)
+							}
+
+							deck = d
+							deck.updateWidgets()
+							return
+						}
+					case error := <-watcher.Errors:
+						fmt.Printf("Watcher had an error: %s\n", error)
```
Comment (muesli) [inline deck.go:100]: I know I've just introduced that in a commit a few minutes ago, but we should try to write all errors to `os.Stderr`:

```go
fmt.Fprintf(os.Stderr, ....)
```

File: deck.go [hunk 4]
```diff
@@ -70,6 +72,42 @@ func LoadDeck(dev *streamdeck.Device, base string, deck string) (*Deck, error) {
...
+					}
+				}
+			}()
+		} else {
+			fmt.Printf("Failed to watch deck, automatic reloading diabled: %s\n", err)
```
Comment (muesli) [inline deck.go:105]: As above, errors to `os.Stderr`.

File: deck.go [hunk 5]
```diff
@@ -70,6 +72,42 @@ func LoadDeck(dev *streamdeck.Device, base string, deck string) (*Deck, error) {
...
+		}
+	} else {
+		fmt.Printf("Failed to initialize fsnotify, automatic reloading diabled: %s\n", err)
```
Comment (muesli) [inline deck.go:108]: As above, errors to `os.Stderr`.

File: main.go [hunk 6]
```diff
@@ -18,7 +18,8 @@ import (
 )
 
 var (
-	deck *Deck
+	deck        *Deck
+	currentDeck string
```
Comment (muesli) [inline main.go:22]: We could probably use `deck.File` here, instead of another global variable.

**Review by muesli — CHANGES_REQUESTED** [review]
Thanks for cleaning this up! Left a few more remarks.

**Comment by muesli** [comment]
A general note: file watching is difficult and I fear there's a bit more work required to make this work universally. Depending on the editor you use to edit a `.deck` file, the file watcher can actually break deckmaster. The reason for that is that some editors don't just overwrite a file, they actually remove the old file and then rename a temporary file to match the name of the edited file. In this case we'll receive an fsnotify event, but loading the deck again will fail as the file has vanished.

Solving this is tricky: you'll have to watch the entire directory the `.deck` file resides in to monitor things like deletions, file creations and move operations.

File: deck.go (hunk 1 above)
Comment (marcusramberg) [reply deck.go:80]: I've updated this so watcher is stored in the deck struct and can be closed before LoadDeck now. It worked from my test as when I'm on a different deck it no longer reverts to main deck when I edit that one.

File: main.go (hunk 6 above)
Comment (marcusramberg) [reply main.go]: Makes sense, done.

**Comment by marcusramberg** [comment]
> A general note: file watching is difficult and I fear there's a bit more work required to make this work universally. Depending on the editor you use to edit a `.deck` file, the file watcher can actually break deckmaster. The reason for that is that some editors don't just overwrite a file, they actually remove the old file and then rename a temporary file to match the name of the edited file. In this case we'll receive an fsnotify event, but loading the deck again will fail as the file has vanished.
> 
> Solving this is tricky: you'll have to watch the entire directory the `.deck` file resides in to monitor things like deletions, file creations and move operations.

I'm not sure which editors this would be a problem for, only been testing on vim and it seems to work very consistently. Did a test autoincrementing a number 250 times (mostly to check that we're not leaking on reload), and seemed to reload consistently every time. An idea to address this issue could be adding a small delay (50-100ms?) (incremented every new event) before saving?

**Comment by marcusramberg** [comment]
I guess there's no interest in progressing this PR so I will just close it.

**Comment by muesli** [comment]
Yeah, to be honest I'm not sure it's worth properly solving this. It would require a bunch of nasty code with tons of (platform and filesystem specific) edge-cases. Thanks for looking into it, tho!

**Pull request closed without merging**